src/ray/rpc/retryable_grpc_client.cc

```diff
@@ -0,0 +1,157 @@
+  // We need to cleanup all the pending requests which are timeout.
+  const auto now = absl::Now();
+  while (!pending_requests_.empty()) {
+    auto iter = pending_requests_.begin();
+    if (iter->first > now) {
+      break;
+    }
+    iter->second->Fail(ray::Status::TimedOut(absl::StrFormat(
+        "Timed out while waiting for %s to become available.", server_name_)));
+    pending_requests_bytes_ -= iter->second->GetRequestBytes();
+    pending_requests_.erase(iter);
+  }
+
+  if (pending_requests_.empty()) {
+    server_unavailable_timeout_time_ = std::nullopt;
+    return;
+  }
+
+  RAY_CHECK(server_unavailable_timeout_time_.has_value());
+
+  auto status = channel_->GetState(false);
+  // https://grpc.github.io/grpc/core/md_doc_connectivity-semantics-and-api.html
+  // https://grpc.github.io/grpc/core/connectivity__state_8h_source.html
+  if (status != GRPC_CHANNEL_READY) {
+    RAY_LOG(DEBUG) << "GRPC channel status: " << status;
+  }
+
+  switch (status) {
+  case GRPC_CHANNEL_TRANSIENT_FAILURE:
+  case GRPC_CHANNEL_CONNECTING: {
+    if (server_unavailable_timeout_time_ < absl::Now()) {
+      RAY_LOG(WARNING) << server_name_ << " has been unavailable for more than "
+                       << server_unavailable_timeout_seconds_ << " seconds";
+      server_unavailable_timeout_callback_();
+      // Reset the unavailable timeout.
+      server_unavailable_timeout_time_ =
+          absl::Now() + absl::Seconds(server_unavailable_timeout_seconds_);
+    }
+
+    if (reset_timer) {
+      SetupCheckTimer();
+    }
+
+    break;
+  }
+  case GRPC_CHANNEL_SHUTDOWN: {
+    RAY_LOG(FATAL) << "Channel shoud never go to this status.";
+    break;
+  }
+  case GRPC_CHANNEL_READY:
+  case GRPC_CHANNEL_IDLE: {
+    server_unavailable_timeout_time_ = std::nullopt;
+    // Retry the ones queued.
+    while (!pending_requests_.empty()) {
+      pending_requests_.begin()->second->CallMethod();
+      pending_requests_.erase(pending_requests_.begin());
+    }
+    pending_requests_bytes_ = 0;
+    break;
+  }
+  default: {
+    RAY_LOG(FATAL) << "Not covered status: " << status;
+  }
+  }
+}
+
+void RetryableGrpcClient::Retry(std::shared_ptr<RetryableGrpcRequest> request) {
+  // In case of transient network error, we queue the request and these requests
+  // will be executed once network is recovered.
+  auto request_bytes = request->GetRequestBytes();
+  auto self = shared_from_this();
+  if (pending_requests_bytes_ + request_bytes > max_pending_requests_bytes_) {
+    RAY_LOG(WARNING) << "Pending queue for failed request has reached the "
+                     << "limit. Blocking the current thread until network is recovered";
+    if (!server_unavailable_timeout_time_.has_value()) {
+      server_unavailable_timeout_time_ =
+          absl::Now() + absl::Seconds(server_unavailable_timeout_seconds_);
+    }
+    while (server_unavailable_timeout_time_.has_value()) {
+      std::this_thread::sleep_for(
```

Comment:
Question: this is something I haven't figured out.
I see async and sync programming mixed in the codebase; for this type of retry, should we use one-shot timer which invokes the callback `[retry_interval]` later?

Reply:
^ @jjyao Do you know the answer?